src/jit/codegencommon.cpp

```diff
@@ -8958,9 +8958,52 @@ void CodeGen::genFnEpilog(BasicBlock* block)
         {
 #ifdef _TARGET_AMD64_
             // Fast tail call.
-            // Call target = RAX.
-            // Stack walker requires that a register indirect tail call be rex.w prefixed.
-            getEmitter()->emitIns_R(INS_rex_jmp, emitTypeSize(TYP_I_IMPL), REG_RAX);
+            GenTreeCall* call = jmpNode->AsCall();
+
+            gtCallTypes callType = (gtCallTypes)call->gtCallType;
+            CORINFO_METHOD_HANDLE methHnd;
+            GenTree* target = call->gtControlExpr;
+            if (callType == CT_INDIRECT)
+            {
+                target = call->gtCallAddr;
+                methHnd = nullptr;
+            }
+            else
+            {
+                methHnd = call->gtCallMethHnd;
+            }
+
+            if (target != nullptr)
```

Comment:
Might be worth commenting why these two if/then/else's are not fused into one.

My impression is that this is for are cases where we have direct calls but still decide to call indirectly, perhaps to avoid needing to use a jump stub or perhaps because of some runtime requirement?

Reply:
I can make this into one check instead. The reason this is two ifs is because I am trying to check whether `genCallInstruction` would have put the address into rax when the call was encountered, so I just copied the code that handles this decision:

https://github.com/dotnet/coreclr/blob/9c5d040b022c9f721c3e9f47d5641cbaa797784a/src/jit/codegenxarch.cpp#L5468-L5480

This entire thing seems somewhat convoluted to me, with three different fields that the target can be in: `gtControlExpr`, `gtCallAddr` or `gtDirectCallAddress`. Maybe my entire check could just be
`if (call->gtDirectCallAddress == nullptr)`. I'll try this.

Reply:
> I'll try this.

Hmm nope, `gtCallDirectAddress` is in a big union and cannot be checked like this.

> My impression is that this is for are cases where we have direct calls but still decide to call indirectly, perhaps to avoid needing to use a jump stub or perhaps because of some runtime requirement?

Yes, this seems to be the case. Direct calls can still be `IAT_PVALUE` or `IAT_PPVALUE` which require indirection and end up with an expression in `gtControlExpr`. Otherwise `gtDirectCallAddress` is set for a truly direct call.
Indirect calls, on the other hand, end up using `gtCallAddr`. I will add a comment.